In [0]:
!pip install mlflow --quiet
dbutils.library.restartPython()

Note: you may need to restart the kernel using %restart_python or dbutils.library.restartPython() to use updated packages.


In [0]:
import mlflow
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn import datasets
 
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import accuracy_score

In [0]:
dataset = datasets.load_iris()

In [0]:
X = dataset.data
y = dataset.target
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [0]:
mlflow.sklearn.autolog()
mlflow.set_experiment("/Users/carrascoalvarezgonzalo@gmail.com/4-ejercicio-the-irish-solucion")

If you are using MLflow Tracing, you can migrate your traces to Unity Catalog for unlimited storage, fine-grained access controls, and queryability from notebooks, SQL, and dashboards. Learn more: https://docs.databricks.com/aws/en/mlflow3/genai/tracing/migrate-traces-to-uc


<Experiment: artifact_location='dbfs:/databricks/mlflow-tracking/1190624933066749', creation_time=1789323741254, effective_trace_archival_retention=None, experiment_id='1190624933066749', last_update_time=1789378841076, lifecycle_stage='active', name='/Users/carrascoalvarezgonzalo@gmail.com/4-ejercicio-the-irish-solucion', tags={'mlflow.experiment.sourceName': '/Users/carrascoalvarezgonzalo@gmail.com/4-ejercicio-the-irish-solucion',
 'mlflow.experimentType': 'MLFLOW_EXPERIMENT',
 'mlflow.ownerEmail': 'carrascoalvarezgonzalo@gmail.com',
 'mlflow.ownerId': '73226958719147'}, trace_location=None, workspace='default'>

# Manual search

In [0]:
configs = [
    {'max_depth': 3, 'max_features': 2},
    {'max_depth': 5, 'max_features': 3},
    {'max_depth': 10, 'max_features': 4},
    {'max_depth': None, 'max_features': 'sqrt'},
]

results = []
for i, config in enumerate(configs):
    with mlflow.start_run(run_name=f"DT Config {i+1}"):
        dt = DecisionTreeClassifier(**config, random_state=42)
        dt.fit(X_train, y_train)

        accuracy = accuracy_score(y_test, dt.predict(X_test))
        results.append({"Config": config, "Accuracy": accuracy})
        print(f"Config {i+1}: {config} -> Accuracy: {accuracy:.4f}")

print(max(results, key=lambda x: x["Accuracy"]))


2026/09/14 10:02:02 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o667.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Config 1: {'max_depth': 3, 'max_features': 2} -> Accuracy: 0.8684


2026/09/14 10:02:11 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o715.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Config 2: {'max_depth': 5, 'max_features': 3} -> Accuracy: 0.8421


2026/09/14 10:02:18 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o755.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Config 3: {'max_depth': 10, 'max_features': 4} -> Accuracy: 0.8421


2026/09/14 10:02:25 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o795.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Config 4: {'max_depth': None, 'max_features': 'sqrt'} -> Accuracy: 0.9211
{'Config': {'max_depth': None, 'max_features': 'sqrt'}, 'Accuracy': 0.9210526315789473}


The best configuration is max_depth: None and max_features: sqrt, which obtained an accuracy of 0.921

# Grid Search

In [0]:
with mlflow.start_run(run_name="Grid Search - Decision Tree"):

  # GridSearch for hyperparameter tuning
  from sklearn.model_selection import GridSearchCV

  param_grid = {
    'max_depth': [3, 5, 10, None],
    'max_features': [2, 3, 4],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
  }
  dt = DecisionTreeClassifier()
  clf = GridSearchCV(estimator=dt, param_grid=param_grid)
  clf.fit(X_train, y_train)
  print(clf.best_params_)
  print(clf.best_estimator_)

  # Predictions
  y_preds = clf.predict(X_test)
  acc = accuracy_score(y_test, y_preds)

  # Log test metrics
  mlflow.log_metric("test_accuracy", acc)

mlflow.end_run()

2026/09/14 09:42:32 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o11.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumented

{'max_depth': None, 'max_features': 2, 'min_samples_leaf': 4, 'min_samples_split': 2}
DecisionTreeClassifier(max_features=2, min_samples_leaf=4)


# Model comparison

In [0]:
import pandas as pd

from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import (
    accuracy_score, 
    precision_score, 
    recall_score, 
    f1_score,
    confusion_matrix,
    classification_report
)

modelos = {
    "Decision tree": DecisionTreeClassifier(max_depth=10, random_state=42),
    "Random forest": RandomForestClassifier(n_estimators=100, random_state=42),
    "Support vector machine": SVC(C=1.0, kernel="rbf", random_state=42),
    "KNN": KNeighborsClassifier(n_neighbors=5),
    "Logistic Regression": LogisticRegression(C=1.0, random_state=42)
}

mlflow.sklearn.autolog()

results = []
for name, model in modelos.items():
    with mlflow.start_run(run_name=f"{name} - Iris"):
        model.fit(X_train, y_train)
        y_preds = model.predict(X_test)

        acc = accuracy_score(y_test, y_preds)
        f1 = f1_score(y_test, y_preds, average="weighted")
        results.append({
            "Model": name,
            "Accuracy": acc,
            "F1-Score": f1
        })
        print(f"{name}: Accuracy = {acc:.4f}, F1 = {f1:.4f}")

# Visualize results
df = pd.DataFrame(results)
print(df.to_string(index=False))

2026/09/14 09:42:49 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o67.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumented

Decision tree: Accuracy = 0.9211, F1 = 0.9212


2026/09/14 09:42:57 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o115.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Random forest: Accuracy = 0.9474, F1 = 0.9474


2026/09/14 09:43:05 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o155.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Support vector machine: Accuracy = 0.8684, F1 = 0.8677


2026/09/14 09:43:14 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o195.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

KNN: Accuracy = 0.9474, F1 = 0.9474


2026/09/14 09:43:21 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o235.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Logistic Regression: Accuracy = 0.8947, F1 = 0.8947
                 Model  Accuracy  F1-Score
         Decision tree  0.921053  0.921190
         Random forest  0.947368  0.947368
Support vector machine  0.868421  0.867735
                   KNN  0.947368  0.947368
   Logistic Regression  0.894737  0.894737


# Error analysis

In [0]:
import numpy as np

incorrect_indices = np.where(y_preds != y_test)[0]
 
print(f"Number of errors: {len(incorrect_indices)}")
print("Error analysis:")
for idx in incorrect_indices:
    true_class = dataset.target_names[y_test[idx]]
    pred_class = dataset.target_names[y_preds[idx]]
    print(f"  Sample {idx}: Real={true_class}, Prediction={pred_class}")
    print(f"    Attributes: {X_test[idx]}")

Number of errors: 4
Error analysis:
  Sample 4: Real=virginica, Prediction=versicolor
    Attributes: [6.  3.  4.8 1.8]
  Sample 18: Real=versicolor, Prediction=virginica
    Attributes: [6.  2.7 5.1 1.6]
  Sample 21: Real=virginica, Prediction=versicolor
    Attributes: [4.9 2.5 4.5 1.7]
  Sample 31: Real=virginica, Prediction=versicolor
    Attributes: [6.3 2.8 5.1 1.5]


# Dimensionality reduction

In [0]:
from sklearn.decomposition import PCA

# Reduce to 2 dimensions
pca = PCA(n_components=2)
# Fit only on the train set to prevent data leakage
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

with mlflow.start_run(run_name=f"PCA - Decision Tree"):
    dt = DecisionTreeClassifier(max_depth=10, random_state=42)
    dt.fit(X_train_pca, y_train)
    y_preds_pca = dt.predict(X_test_pca)

    acc = accuracy_score(y_test, y_preds_pca)
    print(f"Accuracy: {acc}")



2026/09/14 09:45:33 WARNING mlflow.tracking.context.registry: Encountered unexpected error during resolving tags: An error occurred while calling o427.extraContext. Trace:
py4j.security.Py4JSecurityException: Method public scala.collection.immutable.Map com.databricks.backend.common.rpc.CommandContext.extraContext() is not whitelisted on class class com.databricks.backend.common.rpc.CommandContext
	at py4j.security.WhitelistingPy4JSecurityManager.checkCall(WhitelistingPy4JSecurityManager.java:473)
	at py4j.Gateway.invoke(Gateway.java:305)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.$anonfun$invokeMethod$1(InstrumentedCallCommand.scala:19)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke(InstrumentedPy4jCommand.scala:28)
	at com.databricks.backend.daemon.driver.InstrumentedPy4jCommand.instrumentedInvoke$(InstrumentedPy4jCommand.scala:18)
	at com.databricks.backend.daemon.driver.InstrumentedCallCommand.instrumentedInvoke(Instrumente

Accuracy: 0.8421052631578947
